WalletWasabi/Tor/Socks5/TorSocks5ClientPool.cs

```diff
@@ -0,0 +1,320 @@
+		/// <param name="isolateStream"><c>true</c> value is only available for Tor HTTP client to use a new Tor circuit, <c>false</c> otherwise.</param>
+		/// <param name="cancellationToken">Cancellation token to cancel the asynchronous operation.</param>
+		/// <see cref="HttpRequestException">Caller is supposed to catch this exception so that handling of different <see cref="IHttpClient"/> implementations is correct.</see>
+		public async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, bool isolateStream, CancellationToken cancellationToken = default)
+		{
+			// Connecting to loopback's URIs cannot be done via Tor.
+			if (request.RequestUri!.IsLoopback)
+			{
+				return await ClearnetHttpClient.SendAsync(request, cancellationToken).ConfigureAwait(false);
+			}
+
+			int i = 0;
+			int attemptsNo = 3;
+			IPoolItem? poolItem = null;
+
+			try
+			{
+				do
+				{
+					i++;
+					poolItem = await ObtainPoolItemAsync(request, isolateStream, cancellationToken).ConfigureAwait(false);
+					IPoolItem? itemToDispose = poolItem;
+
+					try
+					{
+						Logger.LogTrace($"['{poolItem}'][Attempt #{i}] About to send request.");
+						HttpResponseMessage response = await SendCoreAsync(poolItem.GetTransportStream(), request, cancellationToken).ConfigureAwait(false);
+
+						// Client works OK, no need to dispose.
+						itemToDispose = null;
+
+						// Let others use the client.
+						PoolItemState state = poolItem.Unreserve();
+						Logger.LogTrace($"['{poolItem}'][Attempt #{i}] Unreserve. State is: '{state}'.");
+
+						TorDoesntWorkSince = null;
+						LatestTorException = null;
+
+						return response;
+					}
+					catch (IOException e)
+					{
+						// NetworkStream may throw IOException.
+						throw new TorConnectionException($"Failed to read/write HTTP(s) request.", e);
+					}
+					catch (TorConnectCommandException e) when (e.RepField == RepField.TtlExpired)
+					{
+						// If we get TTL Expired error then wait and retry again linux often does this.
+						Logger.LogTrace($"['{poolItem}'] TTL exception occurred.", e);
+
+						await Task.Delay(3000, cancellationToken).ConfigureAwait(false);
+
+						if (i == attemptsNo)
+						{
+							Logger.LogDebug($"['{poolItem}'] All {attemptsNo} attempts failed.");
+							throw;
+						}
+					}
+					catch (HttpRequestException e)
+					{
+						Logger.LogTrace($"['{poolItem}'] Request may actually fail because we repeat the action.", e);
+
+						if (i == attemptsNo)
+						{
+							Logger.LogDebug($"['{poolItem}'] All {attemptsNo} attempts failed.");
+							throw;
+						}
+					}
+					catch (SocketException e) when (e.ErrorCode == (int)SocketError.ConnectionRefused)
+					{
+						Logger.LogTrace(e);
+						throw new TorConnectionException("Connection was refused.", e);
+					}
+					catch (Exception e)
+					{
+						Logger.LogTrace(e);
+						throw;
+					}
+					finally
+					{
+						(itemToDispose as IDisposable)?.Dispose();
```

Comment:
No implementations of `IPoolItem` are disposable (`TorPoolItem` implements `Dispose(bool)` and `Dispose()` but it is not `IDisposable`)

Reply:
Yes. That's a bug. Thank you.